### PR TITLE
CLN: move top-level dirs

### PR DESCRIPTION
**Pull request by jreback**
xref #13634

- [x] ``from pandas.types.concat import union_categoricals`` ?

- ``pandas.types`` -> ``pandas.core.typed``
- ``pandas.computation`` -> ``pandas.core.computation``
- ``pandas.sparse`` -> ``pandas.core.sparse``

### CODE REVIEW

**Comment by jreback** [comment]
this was probably the one with the most actual code changes.

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -1000,12 +1000,13 @@ New Behavior:
    In [11]: index.memory_usage(deep=True)
    Out[11]: 260
 
-.. _whatsnew_0200.api_breaking.extensions:
+.. _whatsnew_0200.api_breaking.privacy:
 
-Extension Modules Moved
-^^^^^^^^^^^^^^^^^^^^^^^
+Modules Privacy Has Changed
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: we can make this more prominent later.

**Comment by jreback** [comment]
``from pandas.types.concat import union_categoricals``
we need to change this.

``from pandas.api import union_categoricals``?

also should change ``infer_dtype`` to come from ``pandas.api`` as well? (xref #15936) (I will do as a followup I think).

**Comment by jreback** [comment]
cc @chris-b1 @jorisvandenbossche

**Comment by chris-b1** [comment]
I suppose `union_categoricals` and `infer_dtype` should go in `pandas.api.types`?

**Comment by jreback** [comment]
yep I think that's right place.

**Comment by jorisvandenbossche** [comment]
Is there a reason you use `core.typed` instead of `core.types` ?

**Comment by jreback** [comment]
> Is there a reason you use core.typed instead of core.types ?

yes, unfortunately in python 2 if you do ``import types`` and you are *in*  the ``pandas.core`` directly you get ``pandas.core.types``, rather than the built in module (it get shadowed).

we use this in ``pandas.core.series``. I *can* get around this, but simpler to just rename.

**Comment by jorisvandenbossche** [comment]
Ah, just saw you commit message that mentioned it as well. 
I might prefer 'dtypes' over 'typed', but not that important

**Comment by jreback** [comment]
renamed ``typed`` -> ``dtypes``, so a bit more natural now

``pandas.core.dtypes``

**Comment by jreback** [comment]
finally got this bad boy working. apparently ``pytest``doesn't like nested conftests.py (of course *only* when you are running using ``pytest.main()``; it works just fine for development.

**Comment by jorisvandenbossche** [comment]
Great!

**Pull request closed without merging**